### ISSUE TITLE
`cognitive_complexity` lint includes code expanded from macro in calculated complexity

### ISSUE THREAD

**Issue by KonaeAkira** [rust-lang/rust-clippy#14417]
### Summary

Code expanded from macros should not count towards cognitive complexity. Instead, a macro call should have the same cognitive complexity as a function call.

### Lint Name

cognitive_complexity

### Reproducer

I tried this code:

```rust
#![warn(clippy::cognitive_complexity)]

// exact content of the macro is not important
macro_rules! create_4d_vec {
    ( $dim_1:literal, $dim_2:literal, $dim_3:literal, $dim_4:literal) => {
        let mut vec = vec![vec![vec![vec![0; $dim_4]; $dim_3]; $dim_2]; $dim_1];
        for a in 0..$dim_1 {
            for b in 0..$dim_2 {
                for c in 0..$dim_3 {
                    for d in 0..$dim_4 {
                        vec[a][b][c][d] = a + b + c + d;
                    }
                }
            }
        }
    };
}

fn foo() {
    create_4d_vec!(1, 2, 3, 4);
    create_4d_vec!(1, 2, 3, 4);
    create_4d_vec!(1, 2, 3, 4);
    create_4d_vec!(1, 2, 3, 4);
    create_4d_vec!(1, 2, 3, 4);
    create_4d_vec!(1, 2, 3, 4);
    create_4d_vec!(1, 2, 3, 4);
}
```

I saw this happen:

```
warning: the function has a cognitive complexity of (29/25)
  --> src/main.rs:18:4
   |
18 | fn foo() {
   |    ^^^
   |
   = help: you could split it up into multiple smaller functions
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cognitive_complexity
```

I expected to see this happen:

Lint does not trigger.

### Version

```text
rustc 1.87.0-nightly (ecade534c 2025-03-14)
binary: rustc
commit-hash: ecade534c66478c51c5d3c1d3682dc4beb0ac972
commit-date: 2025-03-14
host: x86_64-unknown-linux-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0
```

### Additional Labels

@rustbot label +L-nursery

**Comment by sivizius** [comment]
Exactly the code from https://github.com/rust-lang/rust-clippy/issues/7905 copy-pasted into my code triggers `error: the function has a cognitive complexity of (92/25)`.

Versions:

```shell
$ rustc -Vv                                                                           
rustc 1.88.0 (6b00bc388 2025-06-23) (built from a source tarball)
binary: rustc
commit-hash: 6b00bc3880198600130e1cf62b8f8a93494488cc
commit-date: 2025-06-23
host: x86_64-unknown-linux-gnu
release: 1.88.0
LLVM version: 20.1.6
$ cargo clippy --version --verbose   
clippy 0.1.88
```

But not on the rust playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=ebeb45881388e5af434464e181b0799b

It didn’t triggered in 1.87.0. In fact, it even ignored a lot of very complex functions, afaics those with tracing-macros. Each `info!` seems to increase the complexity by 7, the expanded coded both by the playground and locally expanded looks the same.

I will investigate further why it triggered in my code but not in the playground.

**Comment by sivizius** [comment]
> It didn’t triggered in 1.87.0.

It did trigger, but not the other functions that only trigger in 1.88.0.

Ok, this is strange:

In 1.87.0, the following code does not trigger:

```rust
async fn _async_with_args(_config: ()) {
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    unimplemented!()
}
```

But these do:

```rust
async fn _async_without_args() {
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    unimplemented!()
}

fn _blocking_with_args(_config: ()) {
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    info!("Hello world!");
    unimplemented!()
}
```

Return type does not seem to have any effect.
In 1.88.0, all of them trigger on my PC.

**Comment by sivizius** [comment]
In 1.87.0, this will only trigger, if `tracing.features = [ "log" ]`:

```rust
#[warn(clippy::cognitive_complexity)]
fn _trigger_clippy_cognitive_complexity() {
    log::info!("huh?");
    tracing::info!("Each of these");
    tracing::info!("tracing::info lines");
    tracing::info!("adds 7");
    tracing::info!("cognitive complexity");
    unimplemented!()
}
```

It will not trigger if `_trigger_clippy_cognitive_complexity` is an `async` function with at least one argument.

In 1.88.0 it will trigger even for `async fn` with one or more argument, so, this was a false-negative before.

`tracing` with `log` feature seem to expand to some complex code that does increase the complexity by 7.
